File: runtime/src/main/java/com/epam/drill/auto/test/agent/penetration/junit/JUnitRunnerPenetration.java
package com.epam.drill.auto.test.agent.penetration.junit;

import com.epam.drill.auto.test.agent.AgentClassTransformer;
import com.epam.drill.auto.test.agent.penetration.Strategy;
import javassist.CannotCompileException;
import javassist.CtClass;
import javassist.CtMethod;
import javassist.NotFoundException;

import java.io.IOException;


@SuppressWarnings("all")
public class JUnitRunnerPenetration extends Strategy {


    @Override
    public String id() {
        return "junit-runner";
    }

    public boolean permit(CtClass ctClass) {
        try {
            CtClass[] interfaces = ctClass.getInterfaces();
            for (CtClass in : interfaces) {
                if (in.getName().equals("org.junit.platform.engine.support.hierarchical.Node$DynamicTestExecutor")) {
                    return true;
                }
            }
        } catch (NotFoundException e) {
            e.printStackTrace();
        }
        return false;
    }

    public byte[] instrument(CtClass ctClass) throws CannotCompileException, IOException, NotFoundException {
        CtMethod runChild = ctClass.getDeclaredMethod("execute");
        runChild.insertBefore("String drillTestName = $1.getDisplayName();\n" +
                AgentClassTransformer.CLASS_NAME + ".memorizeTestName(drillTestName);\n"
        );
        return ctClass.toBytecode();
    }

}

File: runtime/src/main/kotlin/com/epam/drill/auto/test/agent/penetration/testing/junit/JUnitRunnerPenetration.kt
package com.epam.drill.auto.test.agent.penetration.testing.junit

import com.epam.drill.auto.test.agent.AgentClassTransformer
import com.epam.drill.auto.test.agent.penetration.AbstractTestStrategy
import javassist.CtClass

class JUnitRunnerPenetration : AbstractTestStrategy() {
    override val id: String = "junit-runner"

    override fun permit(ctClass: CtClass): Boolean {
        return ctClass.interfaces.any { it.name == "org.junit.platform.engine.support.hierarchical.Node\$DynamicTestExecutor" }
    }

    override fun instrument(ctClass: CtClass): ByteArray? {
        val runChild = ctClass.getDeclaredMethod("execute")
        runChild.insertBefore(
            """
                String drillTestName = $1.getDisplayName();
                ${AgentClassTransformer.CLASS_NAME}.INSTANCE.memorizeTestName(drillTestName);
                
                """.trimIndent()
        )
        return ctClass.toBytecode()
    }
}